pkg/compute/assemblyscript.go

```diff
@@ -0,0 +1,230 @@
+	// the binary exists on the users $PATH and error with installation help text.
+	fmt.Fprintf(out, "Checking if npm is installed...\n")
+
+	p, err := exec.LookPath("npm")
+	if err != nil {
+		return errors.RemediationError{
+			Inner:       fmt.Errorf("`npm` not found in $PATH"),
+			Remediation: fmt.Sprintf("To fix this error, install Node.sj and npm by visiting:\n\n\t$ %s", text.Bold("https://nodejs.org/")),
+		}
+	}
+
+	fmt.Fprintf(out, "Found npm at %s\n", p)
+
+	// 2) Check package.json file exists in $PWD
+	//
+	// A valid npm package manifest file is needed for the install command to
+	// work. Therefore, we first assert whether one exists in the current $PWD.
+	fpath, err := filepath.Abs("package.json")
+	if err != nil {
+		return fmt.Errorf("getting package.json path: %w", err)
+	}
+
+	if !common.FileExists(fpath) {
+		return errors.RemediationError{
+			Inner:       fmt.Errorf("package.json not found"),
+			Remediation: fmt.Sprintf("To fix this error, run the following command:\n\n\t$ %s", text.Bold("npm init")),
+		}
+	}
+
+	fmt.Fprintf(out, "Found package.json at %s\n", fpath)
+
+	// Call npm install.
+	cmd := common.NewStreamingExec("npm", []string{"install"}, []string{}, false, out)
+	if err := cmd.Exec(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Build implements the Toolchain interface and attempts to compile the package
+// AssemblyScript source to a Wasm binary.
+func (a AssemblyScript) Build(out io.Writer, verbose bool) error {
+	// Check if bin directory exists and create if not.
+	pwd, err := os.Getwd()
+	if err != nil {
+		return fmt.Errorf("error getting current working directory: %w", err)
+	}
+	binDir := filepath.Join(pwd, "bin")
+	if err := common.MakeDirectoryIfNotExists(binDir); err != nil {
+		return fmt.Errorf("error making bin directory: %w", err)
+	}
+
+	npmdir, err := getNpmBinPath()
+	if err != nil {
+		return err
+	}
+
+	args := []string{
+		"src/index.ts",
+		"--binaryFile",
+		filepath.Join(binDir, "main.wasm"),
+		"--optimize",
+		"--noAssert",
+	}
+	if verbose {
+		args = append(args, "--verbose")
+	}
+
+	// Call asc with the build arguments.
+	cmd := common.NewStreamingExec(filepath.Join(npmdir, "asc"), args, []string{}, verbose, out)
+	if err := cmd.Exec(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getNpmBinPath() (string, error) {
+	path, err := exec.Command("npm", "bin").Output()
+	if err != nil {
+		return "", fmt.Errorf("error getting npm bin path: %w", err)
+	}
+	return strings.TrimSpace(string(path)), nil
+}
+
+func checkPackageDependencyExists(name string) bool {
+	// gosec flagged this:
+	// G204 (CWE-78): Subprocess launched with variable
+	// Disabling as the variables come from trusted sources.
+	/* #nosec */
+	cmd := exec.Command("npm", "link", "--json", "--depth", "0", name)
+	if err := cmd.Start(); err != nil {
+		return false
+	}
+	if err := cmd.Wait(); err != nil {
+		return false
+	}
+	return true
```

Comment:
suggestion - I think this could be simplified to
```suggestion
return exec.Command("npm", "link", "--json", "--depth", "0", name).Run() == nil
```

Reply:
I don't know that I'm a fan of condensing into a single line, but maybe
```suggestion
	cmd := exec.Command("npm", "link", "--json", "--depth", "0", name)
return cmd.Run() == nil
```